File: src/go/problems/contains_nearby_duplicate_test.go
// https://leetcode.com/problems/contains-duplicate-ii

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type ContainsNearbyDuplicate struct{}

func TestContainsNearbyDuplicate(t *testing.T) {
	gen := core.TestSuite[ContainsNearbyDuplicate]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 2, 3, 1}).Param(3).Result(true)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 0, 1, 1}).Param(1).Result(true)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1, 2, 3, 1, 2, 3}).Param(2).Result(false)
	}).Run(t)
}

func (ContainsNearbyDuplicate) Solution(nums []int, k int) bool {
	set := map[int]int{}

	for i, num := range nums {
		if _, ok := set[num]; ok {
			if i-set[num] <= k {
				return true
			}
		}

		set[nums[i]] = i
	}

	return false
}

File: src/python/problems/contains_nearby_duplicate_test.py
# https://leetcode.com/problems/contains-duplicate-ii

from core.problem_base import *

class ContainsNearbyDuplicate(ProblemBase):
    def Solution(self, nums: list[int], k: int) -> bool:
        map = {}
        for i, num in enumerate(nums):
            if num in map and i - map[num] <= k:
                return True
            
            map[nums[i]] = i

        return False

if __name__ == '__main__':
    TestGen(ContainsNearbyDuplicate) \
        .Add(lambda tc: tc.Param([1,2,3,1]).Param(3).Result(True)) \
        .Add(lambda tc: tc.Param([1,0,1,1]).Param(1).Result(True)) \
        .Add(lambda tc: tc.Param([1,2,3,1,2,3]).Param(2).Result(False)) \
        .Run()
